core/src/main/java/org/apache/iceberg/rest/HTTPClient.java

```diff
@@ -468,6 +483,19 @@ public Builder uri(String path) {
       return this;
     }
 
+    public Builder withProxy(String hostname, int port) {
+      Preconditions.checkNotNull(hostname, "Invalid hostname for http client proxy: null");
+      this.proxy = new HttpHost(hostname, port);
+      return this;
+    }
+
+    public Builder withProxyCredentialsProvider(CredentialsProvider credentialsProvider) {
+      Preconditions.checkNotNull(
+          credentialsProvider, "Invalid credentials provider for http client proxy: null");
+      this.credentialsProvider = credentialsProvider;
```

Comment:
I think `proxyCredentialsProvider` would be a better name. Also I think when `build` is called there should be a validation that if `proxyCredentialsProvider` is specified there is a proxy (it should not be the case that someone somehow specifies a proxyCredentialsProvider without having a proxy defined, if so build should fail).

Reply:
So I also thought of failing the build in such cases but another option was to simply end up in a no-op if a proxyCredentialsProvider is specified but not the proxyServer. I ended up doing the latter in the private constructor. 

```
    if (proxy != null) {
      if (credsProvider != null) {
        clientBuilder.setDefaultCredentialsProvider(credsProvider);
      }

      clientBuilder.setProxy(proxy);
    }
```

But I'm happy to add an additional notNull check. Will fix the naming. Thanks!

Reply:
Fixed.